### DISCUSSION TITLE
BCAD panels UI

### DISCUSSION THREAD

**Discussion by schoeller** [CEXT-Dan/PyRx#154] [Ideas]
Hi @CEXT-Dan,

BCAD provides panels in their user interface. For me as a user they are *the* approach to add functionality to their ribbon interface. I guess it would be cool to be able to extend the BCAD interface with Python-based modules. I would admire a panel add-on as a testcase and (if one day capable enough) for example a table editor to show and modify alignment info. Functionality is likely to be limited to BCAD only.

How to read info from alignments has been resolved [here](https://github.com/CEXT-Dan/PyRx/blob/c3bd0aa2e061e47b56fa2450f4ece7951ca342a0/PySamples/PyBrxCv/brxsdk/samples/alignmentsample.py#L101).

<img src="https://github.com/user-attachments/assets/f90a088e-2103-4a29-9787-b79f3f621186" width=300>

When studying the SDK (with my limited skills) I figure that BRX25.1.05.0\inc\BrxSpecific\BcUiPanel.h might be central to adding a panel. I figure that [PyBrx](https://github.com/CEXT-Dan/PyRx/blob/c3bd0aa2e061e47b56fa2450f4ece7951ca342a0/PyRxCore/PyBrx.cpp) might be the right place to extend possibly with PyBrxPanel. If this is possible in combination with wx or whether qt is required is above my head. There is further available in BRX25.1.05.0\inc\BrxSpecific\ribbon which from my understanding steers the top ribbon.

<img src="https://github.com/user-attachments/assets/2c6cf015-c350-4391-acf9-61821279ec7f" width=800>

Please evaluate although I understand there might be more important tasks to tackle. 

Best

Seb

**Comment by CEXT-Dan** [comment]
With regards to the ribbon, in AutoCAD, I make ribbon panels with the CUI command. I’ve done it once with BricsCAD, I think it’s possible to save the partial CUI, Icon, and other resources in a zip, load them with your python application.  The benefit of a CUI is, you can add Menu, toolbar, and Quad entries.  Maybe consider using the CUI? To wrap, it’s around 20 or so classes.

I will look at BcUiPanel, BcUiPanelMFC, also not trivial, but I think I can reuse some of the code from Palettes.
See PySamples\wxWidgets\Palette and http://www.theswamp.org/index.php?topic=58384

There’s some caveats, like in the sample (see MyPanel), the interface is a wxPython wx.Panel. The issue is, there’s no dark mode, I kind of fake it with method set_dark_mode. 

To make feature rich GUIs, you will want to invest (time) in a GUI toolkit that will generate XRC (wxWidgets' XML resources) code. There are a couple free ones listed here https://www.wxpython.org/  I use http://www.anthemion.co.uk/dialogblocks/ it seems its been retired.

**Reply by CEXT-Dan** [reply]
[#156]

**Reply by schoeller** [reply]
I'll start my luck w/ [wxUiEditor](https://github.com/KeyWorksRW/wxUiEditor). First bummer:
`<!-- XRC doesn't support any properties for wxGrid. -->`

**Reply by CEXT-Dan** [reply]
dialog blocks does, I'd try the free version of that, I think you can do up 30 controls

**Reply by CEXT-Dan** [reply]
just to add, you don't have to use XRC.
understand how sizers work is important,  I hated them at first, now I think they're the coolest because you can set them to stretch

**Reply by schoeller** [reply]
Advancing on my ventures, but keeps crashing when doing _PYRELOAD. Images seem to be 64x64px
<img src="https://github.com/user-attachments/assets/b4c9af7c-3ff1-44d3-9457-4ab393c27412" width=360>

**Reply by CEXT-Dan** [reply]
yeah, you probably can’t reload since the panel is sharing resources BrisCAD’s main frame

**Reply by CEXT-Dan** [reply]
I added a destroy function, but it still crashes on reload. I didn't try the BRX sample

**Reply by schoeller** [reply]
seems to be working apart from the boolean test. if I manage to shuvel stuff in a dataframe... (the kids love my spare time activities).

```python
from pyrx_imp import Rx, Ge, Gi, Db, Ap, Ed, Gs, Brx

import wx
import wx.grid
import pandas as pd
import numpy as np

# Generate sample data
num_rows = 40
data = {
    'BooleanColumn': np.random.choice([True, False], size=num_rows),
    'NumericColumn': np.random.randn(num_rows),
    'StringColumn': [f'String{i}' for i in range(num_rows)],
    'DateTimeColumn': pd.date_range(start='2023-01-01', periods=num_rows, freq='D')
}

# Create the DataFrame
df = pd.DataFrame(data)

class DataPacket:
    def __init__(self, message):
        self.message = message

    def classWorker(self, args = None):
        print(args.message)

class MyPanel(wx.Panel):
    def __init__(self):
        super().__init__()
        self.Bind(wx.EVT_SHOW, self.OnShow)
        self.packet = DataPacket("Woohoo!")

    def OnShow(self, event):
        self.choices = df.iloc[:, 2].unique() # Choosing second column for choices
        self.choice = wx.Choice(self, choices=self.choices)
        self.choice.SetSelection(0)

        self.grid = wx.grid.Grid(self)
        self.grid.CreateGrid(df.shape[0], df.shape[1])
        for i in range(df.shape[0]):
            for j in range(df.shape[1]):
                self.grid.SetCellValue(i, j, str(df.iloc[i, j]))

        self.grid.AutoSize()
        self.grid.AutoSizeRows()

        sizer = wx.BoxSizer(wx.VERTICAL)
        sizer.Add(self.choice, 0, wx.EXPAND | wx.TOP | wx.BOTTOM, 4)
        sizer.Add(self.grid, 0, wx.EXPAND | wx.TOP | wx.BOTTOM, 4)
        self.SetSizerAndFit(sizer)
        self.Layout()

        self.Bind(wx.grid.EVT_GRID_CMD_CELL_CHANGED, self.OnCellChange, self.grid)

        self.set_dark_mode(self)

    def OnCellChange(self, event):
        row = event.GetRow()
        col = event.GetCol()
        value = self.grid.GetCellValue(row, col)
        type = df.dtypes.iloc[col]


        try:
            if pd.api.types.is_bool_dtype(type) and self.is_bool_string(value):
                converted_value = bool(value)
                print("boolean")
            elif pd.api.types.is_datetime64_any_dtype(type) and pd.to_datetime(value):
                converted_value = pd.to_datetime(value)
                print("datetime")
            elif pd.api.types.is_numeric_dtype(type) and pd.to_numeric(value):
                converted_value = pd.to_numeric(value)
                print("numeric")
            elif pd.api.types.is_string_dtype(type) and str(value):
                converted_value = str(value)
                print("string")
            else:
                raise(ValueError)
            df.iloc[row, col] = converted_value
            print("value set")

        except (ValueError, TypeError):
            # If conversion fails, update the grid with the DataFrame value
            self.grid.SetCellValue(row, col, str(df.iloc[row, col]))
            print("value reverted")

        print("OnCellChange")

    def is_bool_string(s):
        return str(s).lower() in ['true', 'false', 'yes', 'no', '1', '0']

    def set_dark_mode(self, control):
        bkclr = wx.Colour(brxPanel.backgroundColor())
        fgcolor = wx.Colour(brxPanel.tabTextColor())
        for child in control.GetChildren():
            child.SetForegroundColour(fgcolor)
            child.SetBackgroundColour(bkclr)
            self.set_dark_mode(child)

brxPanel = Brx.PyBrxPanel("My Custom Table Panel", "MyTablePanel")

def createpanel():
    try:
        panel = MyPanel()
        brxPanel.create(panel)
        brxPanel.setIcon("./brxpanel_table.png")
        print(brxPanel.isFloating())
    except Exception as err:
        print(err)

def PyRxCmd_brxpanel_table():
    try:
        createpanel()
    except Exception as err:
        print(err)
```

**Reply by CEXT-Dan** [reply]
Excellent! An you didn’t need to use an XRC. BTW, you can probably filter out controls that look funny in my dark mode hack
